test-data/unit/check-protocols.test

```diff
@@ -0,0 +1,1798 @@
+# flags: --strict-optional
+from typing import Protocol
+
+class MyHashable(Protocol):
+    def __my_hash__(self) -> int:
+        return 0
+
+class C(MyHashable):
+    __my_hash__ = None  # E: Incompatible types in assignment \
+(expression has type None, base class "MyHashable" defined the type as Callable[[MyHashable], int])
+[out]
+
+-- Semanal errors in protocol types
+-- --------------------------------
+
+[case testBasicSemanalErrorsInProtocols]
+from typing import Protocol, Generic, TypeVar, Iterable
+
+T = TypeVar('T', covariant=True)
+S = TypeVar('S', covariant=True)
+
+class P1(Protocol[T, T]): # E: Duplicate type variables in Generic[...] or Protocol[...]
+    def meth(self) -> T:
+        pass
+
+class P2(Protocol[T], Protocol[S]): # E: Only single Generic[...] or Protocol[...] can be in bases
+    def meth(self) -> T:
+        pass
+
+class P3(Protocol[T], Generic[S]): # E: Only single Generic[...] or Protocol[...] can be in bases
+    def meth(self) -> T:
+        pass
+
+class P4(Protocol[T]):
+    attr: Iterable[S] # E: Invalid type "__main__.S"
+
+class P5(Iterable[S], Protocol[T]): # E: If Generic[...] or Protocol[...] is present it should list all type variables
+    def meth(self) -> T:
+        pass
+[out]
+
+[case testProhibitSelfDefinitionInProtocols]
+from typing import Protocol
+
+class P(Protocol):
+    def __init__(self, a: int) -> None:
+        self.a = a # E: Protocol members cannot be defined via assignment to self \
+                   # E: "P" has no attribute "a"
+
+class B: pass
+class C:
+    def __init__(self, a: int) -> None:
+        pass
+
+x: P
+x = B()
+# The above has an incompatible __init__, but mypy ignores this for nominal subtypes?
+x = C(1)
+
+class P2(Protocol):
+    a: int
+    def __init__(self) -> None:
+        self.a = 1
+
+class B2(P2):
+    a: int
+
+x2: P2 = B2()  # OK
+[out]
+
+[case testProtocolsInTypingExtensions]
+from typing_extensions import Protocol, runtime
+
+@runtime
+class P(Protocol):
+    def meth(self) -> int:
+        pass
+
+x: object
+if isinstance(x, P):
+    reveal_type(x)  # E: Revealed type is '__main__.P'
+    reveal_type(x.meth())  # E: Revealed type is 'builtins.int'
+
+class C:
+    def meth(self) -> int:
+        pass
+
+z: P = C()
+[builtins fixtures/dict.pyi]
+[out]
+
+[case testProtocolsCannotInheritFromNormal]
+from typing import Protocol
+
+class C: pass
+class D: pass
+
+class P(C, Protocol): # E: All bases of a protocol must be protocols
+    attr: int
+
+class P2(P, D, Protocol): # E: All bases of a protocol must be protocols
+    pass
+
+p: P2
+reveal_type(p.attr) # E: Revealed type is 'builtins.int'
+[out]
+
+-- Generic protocol types
+-- ----------------------
+
+[case testAutomaticProtocolVariance]
+from typing import TypeVar, Protocol
+
+T = TypeVar('T')
+
+# In case of these errors we proceed with declared variance.
+class Pco(Protocol[T]): # E: Invariant type variable 'T' used in protocol where covariant one is expected
```

Comment:
mypy will ask for a covariant variable, I added a test (we don't have bivariance).